Add support for hwloc 2.0+

SteVwonder opened the pull request:
This PR adds support for hwloc versions 2.0+.  Specifically, it:
- Adds an `#ifdef` block similar to the one in flux-core to do filtering of resources
- Replaces how children are accessed with a more forward-compatible method
- Uses the `resource-query` whitelist feature to normalize the resource set tested across hwloc 1 & 2
  - Updates the test input/output accordingly
- Adds a few tweaks to various tests and test scripts

I have it marked as a WIP because `t4004-match-hwloc.t` still has a single failing test at the end.  Oddly enough, the same resource data and jobspec works perfectly fine with `resource-query` but not with `flux-ion-resource` and the full `sched-fluxion-resource` module.  I also haven't verified that things still work 100% in hwloc 1.

EDIT: Closes #656

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -301,11 +301,32 @@ int resource_reader_hwloc_t::unpack_internal (resource_graph_t &g,
+    if (hwloc_topology_set_io_types_filter(topo,
+                                           HWLOC_TYPE_FILTER_KEEP_IMPORTANT)
+        < 0) {
+      m_err_msg += "hwloc_topology_set_io_types_filter";
```
Somehow my other suggestion didn't make it in. We will need `errno` support for these new error branches before the jump: 
```c++
    errno = EINVAL;
```

dongahn commented:
> I have it marked as a WIP because t4004-match-hwloc.t still has a single failing test at the end. 

@SteVwonder: I think I manually reproduced the issue and this was a problem I wanted to look into further. So this came out in an opportune time. 

Before this test is hung, I get the following error message:

> 2020-06-24T18:30:21.552622Z sched-fluxion-resource.err[0]: match_request_cb: match failed due to match error (id=0): Success
2020-06-24T18:30:21.552668Z sched-fluxion-resource.err[0]: match_request_cb: flux_respond_error: Invalid argument

This probably means that the graph created with hwloc2 support doesn't match w/ the underlying job spec. I will look but this could be the case where more hierarchical resource types than the jobspec supported (e.g., socket).

The bigger issue is, though: when an match error is detected, `flux mini run` (or equivalent) should handle it more gracefully. I actually hit this while I was testing PR #675. 

Let me look at a couple of things and make a specific suggestion.

dongahn commented:
The following patch should work for this PR

```diff
diff --git a/t/t4004-match-hwloc.t b/t/t4004-match-hwloc.t
index 3a994ee4..7aca3868 100755
--- a/t/t4004-match-hwloc.t
+++ b/t/t4004-match-hwloc.t
@@ -119,7 +119,8 @@ test_expect_success 'reloading session/hwloc information with test data' '
 '

 test_expect_success 'loading resource module with default resource info source' '
-    load_resource subsystems=containment policy=high
+    load_resource subsystems=containment policy=high \
+load-whitelist=node,socket,core
 '

 test_expect_success 'match-allocate works with four two-socket jobspecs' '
```

My guess is without this option the resulting graph contains other intermediate resource types between `node` and `socket` such a way that the given jobspec won't match.

```yaml
version: 1
resources:
  - type: node
    count: 1
    with:
      - type: slot
        count: 2
        label: default
        with:
          - type: socket
            count: 1
```

I will push this change as a "fixup" commit to your branch to see if that makes CI happy.

For the "match error" leading to a front end hang issue, I will open up a Issue ticket to address in a separate PR.

dongahn commented:
FYI -- pushed this change.

dongahn commented:
@SteVwonder: the CI is happy now. Once you address the review comments and merge my fixup commit, this PR should be good to be merged.

SteVwonder commented:
Thanks @dongahn for figuring that out!  I appreciate the help!  That reminds me that I should add the latest ubuntu image that was added to flux-core so that we have CI coverage of hwloc 2.X (the current testers only have 1.X).

I'll address your review comments today as well.

dongahn commented:
Awesome! Thanks @SteVwonder.

SteVwonder commented:
I just restarted the Ubuntu 20.04 builder now that the flux-core image is being published (thanks @grondo !!!!).  Looks like both issues occur there (the failed test and the empty `R` in match responses):

```
ok 11 - loading resource module with default resource info source
loading resource module with default resource info source
expecting success: 
    flux ion-resource match allocate ${jobspec_2socket} &&
    flux ion-resource match allocate ${jobspec_2socket} &&
    flux ion-resource match allocate ${jobspec_2socket} &&
    flux ion-resource match allocate ${jobspec_2socket}
JOBID                STATUS               AT                   OVERHEAD (Secs)     
0                    ALLOCATED            2020-06-26T04:22:41  0.0005309581756591797
================================================================================
MATCHED RESOURCES:
JOBID                STATUS               AT                   OVERHEAD (Secs)     
1                    ALLOCATED            2020-06-26T04:22:41  0.00048804283142089844
================================================================================
MATCHED RESOURCES:
JOBID                STATUS               AT                   OVERHEAD (Secs)     
2                    ALLOCATED            2020-06-26T04:22:41  0.00046181678771972656
================================================================================
MATCHED RESOURCES:
JOBID                STATUS               AT                   OVERHEAD (Secs)     
3                    ALLOCATED            2020-06-26T04:22:41  0.0004439353942871094
================================================================================
MATCHED RESOURCES:

```

```
expecting success: 
    remove_resource &&
    flux hwloc reload ${excl_4N4B_sierra2} &&
    load_resource subsystems=containment \
policy=high load-whitelist=node,socket,core,gpu &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    flux ion-resource match allocate ${jobspec_1socket_2gpu} &&
    test_expect_code 16 flux ion-resource match allocate ${jobspec_1socket_2gpu}
OSError: error(16): Device or resource busy
not ok 14 - reloading sierra xml and match allocate
```

I'll begin poking at that tomorrow and next week.

dongahn commented:
@SteVwonder: LMK if you need any help to finalize this. This should def. go into our end-of-July freeze.

SteVwonder commented:
Travis is finally green! :partying_face: https://travis-ci.org/github/flux-framework/flux-sched/builds/708137964

@dongahn, I addressed your comments as a fixup commit.  Let me know if the changes look good, and then I'll squash and this PR should be good to go.

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -309,23 +309,26 @@ int resource_reader_hwloc_t::unpack_internal (resource_graph_t &g,
 #else
-    if (hwloc_topology_set_io_types_filter(topo,
+    if (hwloc_topology_set_io_types_filter (topo,
                                            HWLOC_TYPE_FILTER_KEEP_IMPORTANT)
```
Super minor knit: after adding the whitespace, the two lines are not aligned.

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -309,23 +309,26 @@ int resource_reader_hwloc_t::unpack_internal (resource_graph_t &g,
     }
-    if (hwloc_topology_set_cache_types_filter(topo,
+    if (hwloc_topology_set_cache_types_filter (topo,
                                               HWLOC_TYPE_FILTER_KEEP_STRUCTURE)
```
Same

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -309,23 +309,26 @@ int resource_reader_hwloc_t::unpack_internal (resource_graph_t &g,
     }
-    if (hwloc_topology_set_icache_types_filter(topo,
+    if (hwloc_topology_set_icache_types_filter (topo,
                                                HWLOC_TYPE_FILTER_KEEP_STRUCTURE)
```
Same.

SteVwonder commented:
Thanks @dongahn! I'll address those right away.

I forgot to call out that I did have to add a new commit: https://github.com/flux-framework/flux-sched/pull/677/commits/e350c5b6ad67c7724d24a898270a9af96a1e795d.  If you haven't already, that's probably worth looking over.

dongahn commented:
> I forgot to call out that I did have to add a new commit: e350c5b. If you haven't already, that's probably worth looking over.

Yap that commit looks perfect to me and the detailed commit message is awesome! Great findings!

SteVwonder commented:
Thanks @dongahn for the quick reviews!  Just pushed the fixes and squashed.
